Fix property validation issues

mikepizzo opened the pull request:
<!-- markdownlint-disable MD002 MD041 -->

### Issues

*This pull request fixes issue #1008*
*This pull request fixes issue #1009*

### Description

*Supports V6 capability of:
1) Writing special characters in property names if Validations.None is specified, and
2) Writing undefined untyped properties to open types*

### Checklist (Uncheck if it is not completed)

- [x] *Test cases added*
- [x] *Build and test with one-click build and test script passed*

### Additional work necessary

*None.*

xuzhg commented on src/Microsoft.OData.Core/JsonLight/ODataJsonLightPropertySerializer.cs:
```diff
@@ -261,17 +265,9 @@ private bool IsOpenProperty(ODataProperty property)
-            Debug.Assert(
-                this.MessageWriterSettings.ThrowOnUndeclaredPropertyForNonOpenType,
-                "this.MessageWriterSettings.ThrowOnUndeclaredPropertyForNonOpenType");
-            throw new ODataException(ODataErrorStrings.ValidationUtils_PropertyDoesNotExistOnType(this.currentPropertyInfo.PropertyName, this.currentPropertyInfo.MetadataType.OwningType.FullTypeName()));
```
Remove this is a breaking change?

mikepizzo replied:
We do the check before this, in WriteProperty (which is where we should check the property name, not in WriteUntypedValue). There we correctly check to see if the type is no-open.  So this code is only reached after we've verified that that it's valid.

So the only behavioral change is, if the property is undeclared and the type is open we no longer throw the exception for untyped values.  This enables a code path that was previously an error (but shouldn't have been).   

---
In reply to: [154829236](https://github.com/OData/odata.net/pull/1010#discussion_r154829236) [](ancestors = 154829236)

xuzhg commented on src/Microsoft.OData.Core/JsonLight/ODataJsonLightPropertySerializer.cs:
```diff
@@ -183,9 +183,13 @@ private bool IsOpenProperty(ODataProperty property)
             string propertyName = property.Name;
 
-            if (!this.JsonLightOutputContext.PropertyCacheHandler.InResourceSetScope())
+            if (this.JsonLightOutputContext.MessageWriterSettings.Validations != ValidationKinds.None)
```
maybe:
```C#
if ( ... .InResourceSetScope)
{
          if ( .. != ValidationKinds.None)
          { 
                    ValidatePropertyName(...);
          }
          ...
}
else
{
      ...
}
```

mikepizzo replied:
In the else case, PropertyCacheHandler eventually also called ValidatePropertyName. So it was called in both cases, but in weird places. Moving it out of the if statement provided a common place to validate, regardless of whether or not we were in a resourcesetscope (which shouldn't affect property validation).

---
In reply to: [154829506](https://github.com/OData/odata.net/pull/1010#discussion_r154829506) [](ancestors = 154829506)

xuzhg commented on src/Microsoft.OData.Core/PropertyCache.cs:
```diff
@@ -21,8 +21,6 @@ public PropertySerializationInfo GetProperty(string name, string uniqueName, IEd
             PropertySerializationInfo propertyInfo;
             if (!propertyInfoDictionary.TryGetValue(uniqueName, out propertyInfo))
             {
-                WriterValidationUtils.ValidatePropertyName(name);
```
Just remove the validate?  or use the validate setting?

mikepizzo replied:
This is only called from WriteProperty.  By moving the ValidatePropertyName first in WriteProperty, we do the validation (correctly checking whether or not the type is open) before this code ever gets called.  Which is good, since this code doesn't have the context of the defining type to know whether or not it was open.

---
In reply to: [154829612](https://github.com/OData/odata.net/pull/1010#discussion_r154829612) [](ancestors = 154829612)

xuzhg commented on test/FunctionalTests/Microsoft.OData.Core.Tests/IntegrationTests/Writer/JsonLight/ODataJsonLightWriterIntegrationTests.cs:
```diff
@@ -748,7 +790,8 @@ private static string WriteJsonLightEntry(bool isRequest, Uri serviceDocumentUri
             ODataUri odataUri, Action<ODataWriter> writeAction = null, bool isResourceSet = false,
-            EdmModel model = null)
+            EdmModel model = null,
+            bool ValidatePropertyNames = true)
```
maybe it's better to accept "ValidationKinds" enum?

mikepizzo replied:
Good suggestion, but this is internal test code; we can change easily change it later if/when we add tests that handle other ValidationKinds.

---
In reply to: [154829788](https://github.com/OData/odata.net/pull/1010#discussion_r154829788) [](ancestors = 154829788)

mikepizzo commented:
Thanks Sam!

I added replies to your comments.  The change is a bit subtle, since it looks like I removed validations from some cases and added it to others, but by moving the ValidatePropertyNames to the front of WriteProperty it makes sure that the property name is validated in the right situations and with the right context, rather than further down where the context is not available. The only change in functionality should be that undeclared untyped properties on open types should now succeed rather than fail.

The change in the test method to take ValidationKinds enum, rather than ValidatePropertyNames, probably makes sense, except that it implies we handle other types of validations (we don't)  If I had to make other changes I'd probably take this suggestion, but I don't think it's worth updating the PR for since we) we can easily change this (since it's test code) in the future if we want to handle other ValidationKinds.

From: Sam Xu [mailto:notifications@github.com]
Sent: Monday, December 4, 2017 6:06 PM
To: OData/odata.net <odata.net@noreply.github.com>
Cc: Michael Pizzo <mikep@microsoft.com>; Author <author@noreply.github.com>
Subject: Re: [OData/odata.net] Fix property validation issues (#1010)


@xuzhg approved this pull request.

________________________________

In src/Microsoft.OData.Core/JsonLight/ODataJsonLightPropertySerializer.cs<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FOData%2Fodata.net%2Fpull%2F1010%23discussion_r154829236&data=02%7C01%7Cmikep%40microsoft.com%7C76fc8144062b4d79274208d53b84b17a%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636480363470506028&sdata=oM28hfLqx9l%2BVHbyuKRQY77t927IhGIb25Gf%2FHe3nNI%3D&reserved=0>:

> @@ -261,17 +265,9 @@ private bool IsOpenProperty(ODataProperty property)



         private void WriteUntypedValue(ODataUntypedValue untypedValue)

         {

-            if (!this.MessageWriterSettings.ThrowOnUndeclaredPropertyForNonOpenType)

-            {

-                this.JsonWriter.WriteName(this.currentPropertyInfo.WireName);

-                this.jsonLightValueSerializer.WriteUntypedValue(untypedValue);

-                return;

-            }

-

-            Debug.Assert(

-                this.MessageWriterSettings.ThrowOnUndeclaredPropertyForNonOpenType,

-                "this.MessageWriterSettings.ThrowOnUndeclaredPropertyForNonOpenType");

-            throw new ODataException(ODataErrorStrings.ValidationUtils_PropertyDoesNotExistOnType(this.currentPropertyInfo.PropertyName, this.currentPropertyInfo.MetadataType.OwningType.FullTypeName()));

Remove this is a breaking change?

________________________________

In src/Microsoft.OData.Core/JsonLight/ODataJsonLightPropertySerializer.cs<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FOData%2Fodata.net%2Fpull%2F1010%23discussion_r154829506&data=02%7C01%7Cmikep%40microsoft.com%7C76fc8144062b4d79274208d53b84b17a%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636480363470506028&sdata=UGECwDLasLwlhiMJjOEn86Z7outqLpWsaXsTXJ3%2FeKY%3D&reserved=0>:

> @@ -183,9 +183,13 @@ private bool IsOpenProperty(ODataProperty property)



             string propertyName = property.Name;



-            if (!this.JsonLightOutputContext.PropertyCacheHandler.InResourceSetScope())

+            if (this.JsonLightOutputContext.MessageWriterSettings.Validations != ValidationKinds.None)

maybe:

if ( ... .InResourceSetScope)

{

          if ( .. != ValidationKinds.None)

          {

                    ValidatePropertyName(...);

          }

          ...

}

else

{

      ...

}

________________________________

In src/Microsoft.OData.Core/PropertyCache.cs<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FOData%2Fodata.net%2Fpull%2F1010%23discussion_r154829612&data=02%7C01%7Cmikep%40microsoft.com%7C76fc8144062b4d79274208d53b84b17a%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636480363470506028&sdata=tq3lBblm4djrK1pW1dNsrLk3s7W1UGM%2Fh%2F4MWWUt9BY%3D&reserved=0>:

> @@ -21,8 +21,6 @@ public PropertySerializationInfo GetProperty(string name, string uniqueName, IEd

             PropertySerializationInfo propertyInfo;

             if (!propertyInfoDictionary.TryGetValue(uniqueName, out propertyInfo))

             {

-                WriterValidationUtils.ValidatePropertyName(name);

Just remove the validate? or use the validate setting?

________________________________

In test/FunctionalTests/Microsoft.OData.Core.Tests/IntegrationTests/Writer/JsonLight/ODataJsonLightWriterIntegrationTests.cs<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FOData%2Fodata.net%2Fpull%2F1010%23discussion_r154829788&data=02%7C01%7Cmikep%40microsoft.com%7C76fc8144062b4d79274208d53b84b17a%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636480363470506028&sdata=8uJSskYoewT%2FPg8etQGiT9uNz20qBhCiErdLwitL54s%3D&reserved=0>:

> @@ -748,7 +790,8 @@ private static string WriteJsonLightEntry(bool isRequest, Uri serviceDocumentUri

         private static string WriteJsonLightEntry(bool isRequest, Uri serviceDocumentUri, bool specifySet,

             ODataResource odataEntry, IEdmNavigationSource entitySet, IEdmStructuredType resourceType,

             ODataUri odataUri, Action<ODataWriter> writeAction = null, bool isResourceSet = false,

-            EdmModel model = null)

+            EdmModel model = null,

+            bool ValidatePropertyNames = true)

maybe it's better to accept "ValidationKinds" enum?

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FOData%2Fodata.net%2Fpull%2F1010%23pullrequestreview-81050800&data=02%7C01%7Cmikep%40microsoft.com%7C76fc8144062b4d79274208d53b84b17a%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636480363470506028&sdata=2XJTQgsbHLgLTkaxUL4o2ooLz5jvhEVtX9Nj6APXfI8%3D&reserved=0>, or mute the thread<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAIQnXgrr-88m23jRe8qU_ra520z0iissks5s9KT3gaJpZM4Q1maw&data=02%7C01%7Cmikep%40microsoft.com%7C76fc8144062b4d79274208d53b84b17a%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636480363470506028&sdata=LvPp6v6%2F4ZgHepcn788goDpGXb85om4rXXMmx3vcSKU%3D&reserved=0>.

xuzhg commented:
@mikepizzo Do I need to merge it or you will do it by yourself?
